### ISSUE TITLE
client codegen enums with 'name' or 'ordinal' property

### ISSUE THREAD

**Issue by FidasL** [ExpediaGroup/graphql-kotlin#1075]
**Library Version**
`3.7.0`

**Description**
When my GraphQL schema contains Enum with `name` or `ordinal` property name, it is not possible to use graphql-codegen in  my graphql client. 

<details>
  <summary>Example of generated Enum class</summary>
  
```
enum class EnumClass {
    createdAt,
    createdById,
    id,
    name,
    /**
     * This is a default enum value that will be used when attempting to deserialize unknown value.
     */
    @JsonEnumDefaultValue
    __UNKNOWN_VALUE
}
```
</details>

This generated class causes compile error, because Enums cannot have `name` or `ordinal` property. In my opinion, this is an unexpected behavior from two main reasons:
- there are lots of existing schemas with `name` property that cannot be easily modified. This is making your codegen for graphql client unusable in most use-cases. 
- best-practice is to use uppercase properties in Kotlin enums.

**Possible solution**
Using `@JsonProperty` in property definition can solve this problem. In example below, I defined enum properties in uppercase. With `@JsonProperty` annotation this property is exposed in requests in lowercase. 
<details>
  <summary>Example of possible solution</summary>

```
enum class EnumClass {
    @JsonProperty("name") 
    ID,

    @JsonProperty("name") 
    NAME,

    /**
     * This is a default enum value that will be used when attempting to deserialize unknown value.
     */
    @JsonEnumDefaultValue
    __UNKNOWN_VALUE
}
```
</details>

If you decide to use this solution, I can participate on creating PR for gradle codegen plugin. 

**Possible workaround**
We solved this problem easily with `finalizedBy` gradle task, that replaces all `name` properties in generated files with `@JsonProperty("name") NAME,` definition
<details>
  <summary>Example of workaround</summary>
 
 ```
task replaceEnumNames(type: Copy) {
    def oldPackageName = "com.graphql.generated"
    def newPackageName = "com.graphql.fixed"
    def oldJsonImport = "import com.fasterxml.jackson.annotation.JsonEnumDefaultValue"
    def newJsonImport = "import com.fasterxml.jackson.annotation.JsonEnumDefaultValue\nimport com.fasterxml.jackson.annotation.JsonProperty"

    def generatedDir = "$buildDir/generated/source/graphql/main/com/graphql/generated/"
    def fixedDir = "$buildDir/generated/source/graphql/main/com/graphql/fixed/"

    from generatedDir
    into fixedDir

    filter {
        it.replace(oldPackageName, newPackageName)
            .replace(oldJsonImport, newJsonImport)
            .replaceAll(Pattern.compile('^ {4}name,$'), '    @JsonProperty("name") NAME,')
    }
}
```
</details>

-> This is only workaround and not working solution for all edge-cases

**Comment by smyrick** [comment]
@FidasL Are you only using `graphql-kotlin-client`? Or are you using the server packages as well?

If it is the server, there is nothing we can do to fix this for you as the issue is with Kotlin, not our library, that you can not have a enum with a `name` field.

For the client code, we could look into generating the proper serialization markers for restricted names like `name` or `ordinal` as GraphQL schemas may have enums with these values, regardless of the language used on the server.

**Comment by smyrick** [comment]
@dariuszkuc Do we know when generating the different types, the serialization library we are using so we can change `name` to `NAME` with a proper annotation?

**Comment by FidasL** [comment]
@smyrick Thank you for your response!
We are using graphql-kotlin server too, but we implemented enums ourselves using `@JsonProperty` annotation, so fields are exposed to graphql schema in lowercase while the internal implementation is in uppercase. 

As you mentioned, exactly the same implementation could be used in client-generator (maybe with some optional setting in codegen configuration so it will not affect users that don't want to use enum properties in uppercase)

**Comment by baillyjamy** [comment]
Hello, I am using `graphql-kotlin-client` with kotlinx serialization and I have also this issue. 
Actually, I am working with Shopify GraphQL API. Here an exemple of the generated enum:
```kotlin
/**
 * Represents valid selling plan interval.
 */
enum class SellingPlanInterval {
  /**
   * Day interval.
   */
  DAY,

  /**
   * Month interval.
   */
  MONTH,

  /**
   * Week interval.
   */
  WEEK,

  /**
   * Year interval.
   */
  YEAR,

  /**
   * This is a default enum value that will be used when attempting to deserialize unknown value.
   */
  __UNKNOWN_VALUE
}
```
But the result when I execute my mutation:
```
[KotlinXGraphQLError(message=Variable $input of type SubscriptionContractCreateInput! was provided invalid value for contract.billingPolicy.interval (Expected "{\"name\"=>\"MONTH\", \"ordinal\"=>1}" to be one of: DAY, WEEK, MONTH, YEAR)...)]
```
I just want to generate the name in the query variable.

**Comment by dariuszkuc** [comment]
@baillyjamy it looks like a different issue - your error indicates that enum was incorrectly serialized (or their api requires special format for enums). Can you provide a sample project that shows this issue?

**Comment by baillyjamy** [comment]
Sorry I can't, it need special access that I can't give. Shopify api need just a string value for this case but yeah, my enum value is incorrectly serialized.
In my case, I need "Month" but the output is of graphql client is "{\"name\"=>\"MONTH\", \"ordinal\"=>1}"
I am using ktor graph client with kotlinx serialization version 4.0.0-alpha.15.

**Comment by dariuszkuc** [comment]
@baillyjamy this was a bug in how we were serializing requests, it should be fixed in `4.0.0-alpha.16`.
